chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,109 +37,245 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
     CHAINERX_NEVER_REACH();
 }
 
+std::unique_ptr<Chunk> Chunk::Split(size_t bytesize) {
+    CHAINERX_ASSERT(bytesize_ >= bytesize);
+    if (bytesize_ == bytesize) {
+        return nullptr;
+    }
+
+    std::unique_ptr<Chunk> remaining = std::make_unique<Chunk>(ptr(), bytesize, bytesize_ - bytesize);
+    bytesize_ = bytesize;
+
+    if (next_ != nullptr) {
+        remaining->SetNext(next_);
+        remaining->next()->SetPrev(remaining.get());
+    }
+    next_ = remaining.get();
+    remaining->SetPrev(this);
+
+    return remaining;
+}
+
+void Chunk::MergeWithNext() {
+    CHAINERX_ASSERT(next_ != nullptr);
+    bytesize_ += next_->bytesize();
+    if (next_->next() != nullptr) {
+        next_->next()->SetPrev(this);
+    }
+    next_ = next_->next();
+}
+
+// Pushes a chunk into an appropriate free list
+//
+// Not thread-safe
+void MemoryPool::PushIntoFreeList(std::unique_ptr<Chunk> chunk) {
+    FreeList& free_list = free_bins_[chunk->bytesize()];
+    free_list.emplace_back(std::move(chunk));
+}
+
+void MemoryPool::CompactFreeBins(std::map<size_t, FreeList>::iterator it_start, std::map<size_t, FreeList>::iterator it_end) {
```

Comment:
This function is a bit too difficult to understand.
Please write more comments, perhaps before each of the loops and `erase`.